### ISSUE TITLE
cilium leaks memory when running inside a container

### ISSUE THREAD

**Issue by aanm** [cilium/cilium#1807]
Cilium's container increases its memory consumption as reported by `docker stats`

```
$ docker stats 12345 --no-stream
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O             BLOCK I/O           PIDS
12345              0.96%               534.8 MiB / X.XX GiB   1.66%               0 B / 0 B           0 B / 40.96 kB      40
$ # Roughly 1 hour later
$ docker stats 12345 --no-stream
CONTAINER           CPU %               MEM USAGE / LIMIT       MEM %               NET I/O             BLOCK I/O           PIDS
12345              5.64%               549.6 MiB / X.XX GiB   1.71%               0 B / 0 B           0 B / 139.3 kB      42
```

By entering the container I was able to grab some memory statistics for both the monitor and the agent:

The monitor:

```
$ cat /proc/209/status
Name:   cilium-node-mon
Umask:  0022
State:  S (sleeping)
Tgid:   290
Ngid:   0
Pid:    290
PPid:   1
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
FDSize: 64
Groups:
NStgid: 290
NSpid:  290
NSpgid: 1
NSsid:  1
VmPeak:  1362100 kB
VmSize:  1335636 kB
VmLck:         0 kB
VmPin:    350320 kB
VmHWM:    473368 kB
VmRSS:    473368 kB
RssAnon:          465272 kB
RssFile:            8096 kB
RssShmem:              0 kB
VmData:   623372 kB
VmStk:       132 kB
VmExe:      5476 kB
VmLib:      2048 kB
VmPTE:      1084 kB
VmPMD:        28 kB
VmSwap:        0 kB
HugetlbPages:          0 kB
Threads:        20
SigQ:   0/128578
SigPnd: 0000000000000000
ShdPnd: 0000000000000000
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: ffffffffffc1feff
CapInh: 0000003fffffffff
CapPrm: 0000003fffffffff
CapEff: 0000003fffffffff
CapBnd: 0000003fffffffff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Cpus_allowed:   00ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        13876107
nonvoluntary_ctxt_switches:     2144
$
$
$ # Roughly 1 hour later
$
$ cat /proc/290/status
Name:   cilium-node-mon
Umask:  0022
State:  S (sleeping)
Tgid:   290
Ngid:   0
Pid:    290
PPid:   1
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
FDSize: 64
Groups:
NStgid: 290
NSpid:  290
NSpgid: 1
NSsid:  1
VmPeak:  1362100 kB
VmSize:  1336428 kB
VmLck:         0 kB
VmPin:    350320 kB
VmHWM:    482392 kB
VmRSS:    482392 kB
RssAnon:          474296 kB
RssFile:            8096 kB
RssShmem:              0 kB
VmData:   632756 kB
VmStk:       132 kB
VmExe:      5476 kB
VmLib:      2048 kB
VmPTE:      1096 kB
VmPMD:        28 kB
VmSwap:        0 kB
HugetlbPages:          0 kB
Threads:        20
SigQ:   0/128578
SigPnd: 0000000000000000
ShdPnd: 0000000000000000
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: ffffffffffc1feff
CapInh: 0000003fffffffff
CapPrm: 0000003fffffffff
CapEff: 0000003fffffffff
CapBnd: 0000003fffffffff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Cpus_allowed:   00ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        14132080
nonvoluntary_ctxt_switches:     2202
```

The agent:

```
$ cat 1/status
Name:   cilium-agent
Umask:  0022
State:  S (sleeping)
Tgid:   1
Ngid:   0
Pid:    1
PPid:   0
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
FDSize: 64
Groups:  
NStgid: 1
NSpid:  1
NSpgid: 1
NSsid:  1
VmPeak:  1283844 kB
VmSize:  1283844 kB
VmLck:         0 kB
VmPin:         0 kB
VmHWM:     57076 kB
VmRSS:     54220 kB
RssAnon:           23652 kB
RssFile:           30568 kB
RssShmem:              0 kB
VmData:   185328 kB
VmStk:       132 kB
VmExe:     37268 kB
VmLib:      2256 kB
VmPTE:       320 kB
VmPMD:        24 kB
VmSwap:        0 kB
HugetlbPages:          0 kB
Threads:        20
SigQ:   0/128578
SigPnd: 0000000000000000
ShdPnd: 0000000000000000
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: ffffffffffc1feff
CapInh: 0000003fffffffff
CapPrm: 0000003fffffffff
CapEff: 0000003fffffffff
CapBnd: 0000003fffffffff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Cpus_allowed:   00ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        977308
nonvoluntary_ctxt_switches:     1633
$
$
$ # Roughly 1 hour later
$ cat /proc/1/status 
Name:   cilium-agent
Umask:  0022
State:  S (sleeping)
Tgid:   1
Ngid:   0
Pid:    1
PPid:   0
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
FDSize: 64
Groups:  
NStgid: 1
NSpid:  1
NSpgid: 1
NSsid:  1
VmPeak:  1564492 kB
VmSize:  1564492 kB
VmLck:         0 kB
VmPin:         0 kB
VmHWM:     61208 kB
VmRSS:     59156 kB
RssAnon:           28460 kB
RssFile:           30696 kB
RssShmem:              0 kB
VmData:   204352 kB
VmStk:       132 kB
VmExe:     37268 kB
VmLib:      2256 kB
VmPTE:       348 kB
VmPMD:        24 kB
VmSwap:        0 kB
HugetlbPages:          0 kB
Threads:        22
SigQ:   0/128578
SigPnd: 0000000000000000
ShdPnd: 0000000000000000
SigBlk: 0000000000000000
SigIgn: 0000000000000000
SigCgt: ffffffffffc1feff
CapInh: 0000003fffffffff
CapPrm: 0000003fffffffff
CapEff: 0000003fffffffff
CapBnd: 0000003fffffffff
CapAmb: 0000000000000000
NoNewPrivs:     0
Seccomp:        0
Cpus_allowed:   00ff
Cpus_allowed_list:      0-7
Mems_allowed:   00000000,00000000,00000000,00000001
Mems_allowed_list:      0
voluntary_ctxt_switches:        985224
nonvoluntary_ctxt_switches:     1721
````

The monitor seems to have a higher rate of memory consumption than the agent.

I've tested cilium (**not the same version running in the container**) without running in a container and the memory consumptions seems stable.

**Comment by manalibhutiyani** [comment]
Are you looking at VmPeak, VmSize and VMData? It looks like the agent is consuming more as per your logs?

**Comment by tgraf** [comment]
Fixed by #1821
